contrib/cloud.google.com/go/pubsub.v1/pubsub.go

```diff
@@ -0,0 +1,87 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// Package pubsub provides functions to trace the cloud.google.com/pubsub/go package.
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	"cloud.google.com/go/pubsub"
+)
+
+// Publish publishes a message on the specified topic and returns a PublishResult.
+// This function is functionally equivalent to t.Publish(ctx, msg), but it also starts a publish
+// span and it ensures that the datadog tracing metadata is propagated as attributes attached to
+// the published message.
+func Publish(ctx context.Context, t *pubsub.Topic, msg *pubsub.Message) *PublishResult {
+	span, ctx := tracer.StartSpanFromContext(
+		ctx,
+		"pubsub.publish",
+		tracer.ResourceName(t.String()),
+		tracer.SpanType(ext.SpanTypeMessageProducer),
+		tracer.Tag("message_size", len(msg.Data)),
+		tracer.Tag("ordering_key", msg.OrderingKey),
+	)
+	if msg.Attributes == nil {
+		msg.Attributes = make(map[string]string)
+	}
+	if err := tracer.Inject(span.Context(), tracer.TextMapCarrier(msg.Attributes)); err != nil {
+		log.Debug("contrib/cloud.google.com/go/pubsub.v1/: failed injecting tracing attributes: %v", err)
+	}
+	span.SetTag("num_attributes", len(msg.Attributes))
+	return &PublishResult{
+		PublishResult: t.Publish(ctx, msg),
+		span:          span,
+	}
+}
+
+// PublishResult wraps *pubsub.PublishResult
+type PublishResult struct {
+	*pubsub.PublishResult
+	once sync.Once
+	span tracer.Span
+}
+
+// Get wraps (pubsub.PublishResult).Get(ctx). When this function returns the publish span
+// created in Publish is completed.
+func (r *PublishResult) Get(ctx context.Context) (string, error) {
+	serverID, err := r.PublishResult.Get(ctx)
+	r.once.Do(func() {
+		r.span.SetTag("server_id", serverID)
+		r.span.Finish(tracer.WithError(err))
+	})
+	return serverID, err
+}
+
+// ReceiveTracer returns a receive callback that wraps the supplied callback, and extracts the datadog tracing metadata
+// if it exists attached to the received message.
+func ReceiveTracer(s *pubsub.Subscription, f func(context.Context, *pubsub.Message)) func(context.Context, *pubsub.Message) {
+	return func(ctx context.Context, msg *pubsub.Message) {
+		parentSpanCtx, _ := tracer.Extract(tracer.TextMapCarrier(msg.Attributes))
+		span, ctx := tracer.StartSpanFromContext(
+			ctx,
```

Comment:
Just so you're aware, the only purpose of this parameter is to extract the span from `ctx` as a parent. If both `ctx` and `tracer.ChildOf` are provided, the latter takes precedence. The code is fine, but I'm wondering if it would be better and cleaner to just leave out `ctx` completely to avoid any confusion to the reader. You don't have to make this change but I feel it would be nicer if we would illustrate this behaviour by using `context.Background()` here and using `_` for the context name in the first function argument.

That would then make it clear that the `context.Context` is never used and has no influence on behaviour, which is the case anyway.

But I might be missing something.

Reply:
Sorry, I missed that `ctx` is used down below again. Please ignore this comment.

Reply:
Also one other thing: `ctx` may contain other things, e.g. loggers or other stuff needed by something else down the Publish call stack (e.g. gRPC middleware, that we use extensively). If I did pass context.Background() to StartSpanFromContext I would lose all these other values.

Reply:
Got it. All good. Resolving.